src/smt/difficulty_post_processor.h

```diff
@@ -0,0 +1,77 @@
+/******************************************************************************
+ * Top contributors (to current version):
+ *   Andrew Reynolds
+ *
+ * This file is part of the cvc5 project.
+ *
+ * Copyright (c) 2009-2021 by the authors listed in the file AUTHORS
+ * in the top-level source directory and their institutional affiliations.
+ * All rights reserved.  See the file COPYING in the top-level source
+ * directory for licensing information.
+ * ****************************************************************************
+ *
+ * Implementation of module for processing the difficulty of input assumptions
+ * based on proof nodes.
+ */
+
+#include "cvc5_private.h"
+
+#ifndef CVC5__SMT__DIFFICULTY_POST_PROCESSOR_H
+#define CVC5__SMT__DIFFICULTY_POST_PROCESSOR_H
+
+#include <map>
+#include <sstream>
+#include <unordered_set>
+
+#include "proof/proof_node_updater.h"
+
+namespace cvc5 {
+
+class Env;
+
+namespace smt {
+
+/**
+ * A postprocess callback that computes the difficulty based on the structure
+ * of the proof. In particular, this class assesses what the source of an
+ * assertion was by considering the shape of the proof. For instance, if
+ * assertion A entails x=t, and this was used to derive a substitution
+ * { x -> t } to assertion B, then B is the source of B*{ x -> t }. The
+ * difficulty of this assertion is carried to B and not A.
+ */
+class DifficultyPostprocessCallback : public ProofNodeUpdaterCallback
+{
+ public:
+  DifficultyPostprocessCallback(Env& env);
+  ~DifficultyPostprocessCallback() {}
+  /**
+   * Set current difficulty of the next proof to process to the (integer)
+   * value stored in Node d.
+   */
+  bool setCurrentDifficulty(Node d);
+  /**
+   * Should proof pn be updated? This is used to selectively traverse to e.g.
+   * the source of an assertion.
+   */
+  bool shouldUpdate(std::shared_ptr<ProofNode> pn,
+                    const std::vector<Node>& fa,
+                    bool& continueUpdate) override;
+  /** Get the (acculumated) difficulty map for the last processed proof node */
+  void getDifficultyMap(std::map<Node, Node>& dmap) const;
+
+ private:
+  /** Reference to the env class */
+  Env& d_env;
+  /**
+   * The current difficulty of the assertion whose proof of preprocessing
+   * we are considering.
```

Comment:
This alludes to a control flow that is not clear.

Reply:
This class assigns a constant difficulty to all free assumptions in the given proof, since it processes the proof of each preprocessed assertion independently. It is the proof manager that ties this together. I'll make this clear in the doc.

Reply:
Cool, thanks.